Change PyTorch tests to use non-default CUDA stream

fbhuba opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#21474 Change PyTorch tests to use non-default CUDA stream**

Summary: Introduce context manager that switches CUDA devices to non-default 
streams on enter and switches back to the original streams on exit. This context 
manager is used to enforce non-default streams in CUDA tests and should be 
used in other tests as well. Pytorch provides required synchronization when default 
streams are used but with non-default streams synchronization is the responsibility 
of the developer. Potential races under non-default streams make it hard to turn it on
on all tests and should rather be done on test by test basis.

Test Plan: Unit tests

Reviewers: ezyang

Subscribers:

Tasks: T40579608

Tags:

Differential Revision: [D15699700](https://our.internmc.facebook.com/intern/diff/D15699700)

Commit:
Change PyTorch tests to use non-default CUDA stream

Summary: Before test starts CUDA memory leak checker replaces
current streams for each CUDA device with new stream and after
test ends restores current streams to their previous values.

Note that some tests rely on a particular CUDA device to be
selected hence it also preserves the currently selected device.

Test Plan: Unit tests

Reviewers: ezyang

Subscribers:

Tasks: T40579608

Tags:

Commit:
Update on "Change PyTorch tests to use non-default CUDA stream"

Change PyTorch tests to use non-default CUDA stream

Summary: Before test starts CUDA memory leak checker replaces
current streams for each CUDA device with new stream and after
test ends restores current streams to their previous values.

Note that some tests rely on a particular CUDA device to be
selected hence it also preserves the currently selected device.

Test Plan: Unit tests

Reviewers: ezyang

Subscribers:

Tasks: T40579608

Tags:

gh-metadata: pytorch pytorch 21474 gh/fbhuba/2/head

Commit:
Update on "Change PyTorch tests to use non-default CUDA stream"

Change PyTorch tests to use non-default CUDA stream

Summary: Before test starts CUDA memory leak checker replaces
current streams for each CUDA device with new stream and after
test ends restores current streams to their previous values.

Note that some tests rely on a particular CUDA device to be
selected hence it also preserves the currently selected device.

Test Plan: Unit tests

Reviewers: ezyang

Subscribers:

Tasks: T40579608

Tags:

gh-metadata: pytorch pytorch 21474 gh/fbhuba/2/head

Commit:
Update on "Change PyTorch tests to use non-default CUDA stream"

Change PyTorch tests to use non-default CUDA stream

Summary: Before test starts CUDA memory leak checker replaces
current streams for each CUDA device with new stream and after
test ends restores current streams to their previous values.

Note that some tests rely on a particular CUDA device to be
selected hence it also preserves the currently selected device.

Test Plan: Unit tests

Reviewers: ezyang

Subscribers:

Tasks: T40579608

Tags:

gh-metadata: pytorch pytorch 21474 gh/fbhuba/2/head

Commit:
Update on "Change PyTorch tests to use non-default CUDA stream"

Change PyTorch tests to use non-default CUDA stream

Summary: Before test starts CUDA memory leak checker replaces
current streams for each CUDA device with new stream and after
test ends restores current streams to their previous values.

Note that some tests rely on a particular CUDA device to be
selected hence it also preserves the currently selected device.

Test Plan: Unit tests

Reviewers: ezyang

Subscribers:

Tasks: T40579608

Tags:

gh-metadata: pytorch pytorch 21474 gh/fbhuba/2/head

Commit:
Update on "Change PyTorch tests to use non-default CUDA stream"

Change PyTorch tests to use non-default CUDA stream

Summary: Before test starts CUDA memory leak checker replaces
current streams for each CUDA device with new stream and after
test ends restores current streams to their previous values.

Note that some tests rely on a particular CUDA device to be
selected hence it also preserves the currently selected device.

Test Plan: Unit tests

Reviewers: ezyang

Subscribers:

Tasks: T40579608

Tags:

gh-metadata: pytorch pytorch 21474 gh/fbhuba/2/head

Commit:
Update on "Change PyTorch tests to use non-default CUDA stream"

Change PyTorch tests to use non-default CUDA stream

Summary: Before test starts CUDA memory leak checker replaces
current streams for each CUDA device with new stream and after
test ends restores current streams to their previous values.

Note that some tests rely on a particular CUDA device to be
selected hence it also preserves the currently selected device.

Test Plan: Unit tests

Reviewers: ezyang

Subscribers:

Tasks: T40579608

Tags:

gh-metadata: pytorch pytorch 21474 gh/fbhuba/2/head

Commit:
Update on "Change PyTorch tests to use non-default CUDA stream"

Change PyTorch tests to use non-default CUDA stream

Summary: Before CUDA test starts CUDA policy replaces
current streams for each CUDA device with new stream and after
test ends restores current streams to their previous values.

Note that some tests rely on a particular CUDA device to be
selected hence it also preserves the currently selected device.

Test Plan: Unit tests

Reviewers: ezyang

Subscribers:

Tasks: T40579608

Tags:

gh-metadata: pytorch pytorch 21474 gh/fbhuba/2/head

ezyang commented:
It looks like you turned it off universally for ~almost everything? I guess that's OK to land for now, I suppose.

ezyang commented:
It would definitely be worth remarking about that in the commit description though!

Commit:
Update on "Change PyTorch tests to use non-default CUDA stream"

Change PyTorch tests to use non-default CUDA stream

Summary: Before CUDA test starts CUDA policy replaces
current streams for each CUDA device with new stream and after
test ends restores current streams to their previous values.

Note that some tests rely on a particular CUDA device to be
selected hence it also preserves the currently selected device.

Test Plan: Unit tests

Reviewers: ezyang

Subscribers:

Tasks: T40579608

Tags:

gh-metadata: pytorch pytorch 21474 gh/fbhuba/2/head
